### DISCUSSION TITLE
Timestamp Precision

### DISCUSSION THREAD

**Discussion by DavidBuchanan314** [bluesky-social/atproto#1550] [Protocol (atproto)]
https://atproto.com/specs/lexicon lists `1985-04-12T23:20:50.1235678912345Z` as a supported timestamp format, but this timestamp is rejected by bsky.social as invalid:

```
Invalid app.bsky.feed.post record: Record/createdAt must be an iso8601 formatted datetime
```

It would appear that, in practice, the maximum supported precision is 9 digits.

A maximum of 9 digits feels sensible, so perhaps the spec/docs should be updated to reflect this?

(as a side-note, python's standard library `datetime` class can only represent microsecond precision, but that's more of a python standard library deficiency than anything else)

**Comment by bnewbold** [comment]
Good catch. It is on my list to revisit the `datetime` string format syntax details in the spec, the typescript reference implementation, and the golang implementation, and to provide interop test files for others to work against. I think the typescript code is currently quite permissive, compared to the spec, and we need to figure out how strict it should be in which contexts.

As some background context (you probably know already), datetime string formats are notoriously finicky for interop in general, and have bitten us several times already times in atproto specifically. We have discussed switching to UNIX integer timestamps, but those also either require somewhat arbitrary precision (millis? micros? whole seconds?) or floats, and we don't really do floats in the atproto data model right now. And we already have a lot of records with formatted strings, so those will probably stick around indefinitely.
